posix/fdtable.c

```diff
@@ -0,0 +1,232 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <openenclave/bits/safecrt.h>
+#include <openenclave/corelibc/errno.h>
+#include <openenclave/corelibc/stdio.h>
+#include <openenclave/corelibc/stdlib.h>
+#include <openenclave/corelibc/string.h>
+#include <openenclave/corelibc/unistd.h>
+#include <openenclave/enclave.h>
+#include <openenclave/internal/posix/device.h>
+#include <openenclave/internal/posix/fdtable.h>
+#include <openenclave/internal/posix/raise.h>
+#include <openenclave/internal/print.h>
+#include <openenclave/internal/thread.h>
+#include <openenclave/internal/trace.h>
+#include <openenclave/internal/utils.h>
+
+/*
+**==============================================================================
+**
+** Define the table of file-descriptor entries:
+**
+**==============================================================================
+*/
+
+/* The table allocation grows in multiples of the chunk size. */
+#define TABLE_CHUNK_SIZE 1024
+
+typedef oe_device_t* entry_t;
+
+static entry_t* _table;
+static size_t _table_size;
+static oe_spinlock_t _lock = OE_SPINLOCK_INITIALIZER;
+static bool _installed_atexit_handler;
+
+static void _atexit_handler(void)
+{
+    oe_free(_table);
+}
+
+static int _resize_table(size_t new_size)
+{
+    int ret = -1;
+
+    /* Install the atexit handler on the first call. */
+    if (!_installed_atexit_handler)
+    {
+        oe_atexit(_atexit_handler);
+        _installed_atexit_handler = true;
+    }
+
+    /* The fdtable cannot be bigger than the maximum int file descriptor. */
+    if (new_size > OE_INT_MAX)
+        goto done;
+
+    /* Round the new capacity up to the next multiple of the chunk size. */
+    new_size = oe_round_up_to_multiple(new_size, TABLE_CHUNK_SIZE);
+
+    if (new_size > OE_INT_MAX)
+        new_size = OE_INT_MAX;
+
+    if (new_size > _table_size)
+    {
+        entry_t* p;
+        size_t n = new_size;
+
+        /* Reallocate the table. */
+        if (!(p = oe_realloc(_table, n * sizeof(entry_t))))
+            goto done;
+
+        /* Zero-fill the unused porition. */
+        {
+            const size_t num_bytes = (n - _table_size) * sizeof(entry_t);
+
+            if (oe_memset_s(p + _table_size, num_bytes, 0, num_bytes) != OE_OK)
+                goto done;
+        }
+
+        _table = p;
+        _table_size = new_size;
+    }
+
+    ret = 0;
+
+done:
+    return ret;
+}
+
+/*
+**==============================================================================
+**
+** Public interface:
+**
+**==============================================================================
+*/
+
+int oe_fdtable_assign(oe_device_t* device)
+{
+    int ret = -1;
+    size_t index;
+
+    oe_spin_lock(&_lock);
+
+    if (!device)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    /* The file-descriptor table must be big enough for standard devices. */
+    if (_table_size < OE_STDERR_FILENO + 1)
```

Comment:
This should check against OE_STDERR_FILENO + 2 since it starts the indexing check on OE_STDERR_FILENO. Granted the off-by-one is trivial since we always allocate at least `TABLE_CHUNK_SIZE` but you might as well avoid the theoretical double `_resize_table` in case of future edits.

It seems like this is much more succinctly handled in `oe_fdtable_reassign` which just default sizes `TABLE_CHUNK_SIZE` up front.

Reply:
Good suggestion. I changed it to use **OE_TABLE_CHUNK_SIZE**. Also, I did not see an off-by-one error. The loop begins at **OE_STDERR_FILENO+1** but it only enters the body if that value is less than **__table_size**.